src/Control/Effect/Carrier.hs

```diff
@@ -1,31 +1,28 @@
-{-# LANGUAGE DefaultSignatures, DeriveFunctor, FlexibleInstances, FunctionalDependencies, RankNTypes, UndecidableInstances #-}
+{-# LANGUAGE DefaultSignatures, DeriveFunctor, EmptyCase, FlexibleContexts, FlexibleInstances, FunctionalDependencies, RankNTypes, TypeOperators, UndecidableInstances #-}
 module Control.Effect.Carrier
 ( HFunctor(..)
 , Effect(..)
 , Carrier(..)
 , handlePure
 , handleCoercible
+-- * Generic deriving of 'HFunctor' & 'Effect' instances.
+, GHFunctor(..)
+, GEffect(..)
 ) where
 
 import Data.Coerce
+import GHC.Generics
 
+-- | Higher-order functors of kind @(* -> *) -> (* -> *)@ map functors to functors.
+--
+--   All effects must be 'HFunctor's.
 class HFunctor h where
-  -- | Functor map. This is required to be 'fmap'.
-  --
-  --   This can go away once we have quantified constraints.
-  fmap' :: (a -> b) -> (h m a -> h m b)
-  default fmap' :: Functor (h m) => (a -> b) -> (h m a -> h m b)
-  fmap' = fmap
-  {-# INLINE fmap' #-}
-
   -- | Higher-order functor map of a natural transformation over higher-order positions within the effect.
-  -- A definition for 'hmap' over first-order effects can be derived automatically.
-  hmap :: (forall x . m x -> n x) -> (h m a -> h n a)
-
-  default hmap :: Coercible (h m a) (h n a)
-               => (forall x . m x -> n x)
-               -> (h m a -> h n a)
-  hmap _ = coerce
+  --
+  -- A definition for 'hmap' over first-order effects can be derived automatically provided a 'Generic1' instance is available.
+  hmap :: Functor m => (forall x . m x -> n x) -> (h m a -> h n a)
```

Comment:
The addition of a `Functor` constraint for `m` means that we admit strictly more instances, e.g.:

```haskell
newtype Scope f a = Scope (f (Maybe (f a)))

instance HFunctor Scope where hmap f (Scope b) = Scope (f (fmap (fmap f) b))
```

would not be possible without the `Functor` constraint on `m`.

On the other hand, it also means that we allow strictly fewer callers, since before you could’ve `hmap`’d to shove a `Set` in there.

I’m ok with this. Use _The Constrained-Monad Problem_ or `Coyoneda` if you want to do that sort of thing.